### DISCUSSION TITLE
Question on hash generation

### DISCUSSION THREAD

**Discussion by GeeWizWow** [vercel/turborepo#789] [Help] [answered]
The documentation for hashing describes that amongst other things, the following is taken into account:
- The outputs option specified in the [pipeline](https://turborepo.org/docs/reference/configuration#pipeline)

Would this not defeat the purpose of remote caching? I.E. would a CI runner having pulled latest, not generate different hashes to a developer who happens to have *output* present on their machine?

Or is the idea that CI would exclusively leverage cache from other CI pipelines runs?

Alternatively would it be considered best practice when using turbo to ensure all *output* is removed in development environments prior to a pipeline run?

Thanks 👋

**Comment by samchouse** [comment, marked as answer]
A pipeline doesn't refer to CI, it's kind of like a plan for turborepo. Eg. if I run build make sure to run lint, then build dependencies, then build target. Remote caching allows you to download the cache from local to CI or any other situation. Eg. if I run build on local machine and have remote caching, it will save locally and be put on a remote server. Then in CI it will download the cache, assuming you have the same hash, that you sent to the remote server from your local machine.

**Reply by GeeWizWow** [reply]
Thanks, so far so good, following your scenario wouldn't the following occur?

I run a build on my local machine, but happen to have an old, gitignored dist folder kicking around at the time of build, given this line in the documentation:

> The outputs option specified in the [pipeline](https://turborepo.org/docs/reference/configuration#pipeline)

the contents of that dist folder are taken into account at hash generation time, that produces 'xyz'.

CI comes along a bit later, starts prepping for a build, generates a hash, which produces `abc` because it doesn't have this old dist folder lying around... hash mistmatch, any previously cached remote artefacts are ignored?

Or am I missing something?

**Reply by samchouse** [reply]
Yeah, it should work like that.

**Reply by GeeWizWow** [reply]
> Yeah, it should work like that.

Cool thanks 🙏 

In that case, just to confirm, when wanting to leverage remote caching, it's especially important to ensure that anything specified in the `outputs` option is removed prior to calling a turbo pipeline?

**Reply by samchouse** [reply]
Kinda, outputs is want you want restored. Eg. Weback creates a bundle in `build`, I would want to add it to outputs. TypeScript compiling and making a build info file, I wouldn't cache that so I wouldn't add it to outputs. Anything that's not in outputs and not a few predefined directories, is used for calculating hash.

**Reply by GeeWizWow** [reply]
Yeah, we're very much in agreement, other than this
> Anything that's not in outputs and not a few predefined directories, is used for calculating hash.

seems to go against what's on this page https://turborepo.org/docs/features/caching#hashing, hence why I'm confused

**Reply by samchouse** [reply]
Whoops, I think outputs is included but some hardcoded folders aren't.
